src/backend/common/RenderPass.cpp

```diff
@@ -0,0 +1,103 @@
+// Copyright 2017 The NXT Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "RenderPass.h"
+
+#include "Buffer.h"
+#include "Device.h"
+#include "Texture.h"
+
+namespace backend {
+
+    // RenderPass
+
+    RenderPassBase::RenderPassBase(RenderPassBuilder* builder)
+        : attachments(std::move(builder->attachments)), subpasses(std::move(builder->subpasses)) {
+    }
+
+    nxt::TextureFormat RenderPassBase::GetAttachmentFormat(uint32_t attachment) const {
+        ASSERT(attachment < attachments.size());
+        return attachments[attachment];
+    }
+
+    uint32_t RenderPassBase::GetSubpassCount() const {
+        return subpasses.size();
+    }
+
+    const RenderPassBase::SubpassInfo& RenderPassBase::GetSubpassInfo(uint32_t subpass) const {
+        ASSERT(subpass < subpasses.size());
+        return subpasses[subpass];
+    }
+
+    // RenderPassBuilder
+
+    enum RenderPassSetProperties {
+        RENDERPASS_PROPERTY_SUBPASS_COUNT = 0x1,
+    };
+
+    RenderPassBuilder::RenderPassBuilder(DeviceBase* device)
+        : device(device), subpasses(1) {
+    }
+
+    bool RenderPassBuilder::WasConsumed() const {
+        return consumed;
+    }
+
+    RenderPassBase* RenderPassBuilder::GetResult() {
+        consumed = true;
```

Comment:
This is sort of a fuzzy concept as these could have optional properties with default values, but I've added this validation for the format field of attachments at least. A subpass by default will just not have any attachments.